File: contracts/spec/motion.ts
// Fused motion state: the contract between a native motion driver and an app.
//
// The native driver owns sampling, calibration and sensor fusion and publishes
// one MotionState per estimate, including the angle conveniences it derives
// from its core members. Applications receive fused state only, never
// accelerometer, gyroscope or magnetometer readings, and the runtime reads the
// state without computing on it.
//
// Vectors use the device frame of W3C DeviceMotionEvent and Android: +x toward
// the right edge, +y toward the top edge, +z out of the screen. World frames
// have +z up, opposite gravity.

/** Fusion depth a driver provides; each level adds state to the previous one. */
export const MotionLevel = { Gravity: 1, Inertial: 2, Geomagnetic: 3 } as const;
export type MotionLevelId = (typeof MotionLevel)[keyof typeof MotionLevel];
export const MOTION_LEVELS = { gravity: MotionLevel.Gravity, inertial: MotionLevel.Inertial, geomagnetic: MotionLevel.Geomagnetic } as const;
export type MotionLevelName = keyof typeof MOTION_LEVELS;

/** Trust in one estimate. `Unavailable` means the value is absent and zeroed. */
export const MotionQuality = { Unavailable: 0, Unreliable: 1, Low: 2, Medium: 3, High: 4 } as const;
export type MotionQualityId = (typeof MotionQuality)[keyof typeof MotionQuality];
/** Thresholds an app may require; an unavailable value never reaches a handler. */
export const MOTION_MIN_QUALITIES = { unreliable: MotionQuality.Unreliable, low: MotionQuality.Low, medium: MotionQuality.Medium, high: MotionQuality.High } as const;
export type MotionMinQualityName = keyof typeof MOTION_MIN_QUALITIES;
export const MOTION_DEFAULT_MIN_QUALITY: MotionMinQualityName = "low";

/**
 * Frame an estimate is expressed in. `Local` has +z up and a horizontal origin
 * fixed when the driver last aligned; `MagneticNorth` and `TrueNorth` are
 * east-north-up frames.
 */
export const MotionReferenceFrame = { Device: 0, Local: 1, MagneticNorth: 2, TrueNorth: 3 } as const;
export type MotionReferenceFrameId = (typeof MotionReferenceFrame)[keyof typeof MotionReferenceFrame];

export interface MotionVectorEstimate { value: readonly [number, number, number]; quality: MotionQualityId }
export interface MotionScalarEstimate { value: number; quality: MotionQualityId }
export interface MotionOrientationEstimate {
  /** Unit quaternion (w, x, y, z) rotating device-frame vectors into `referenceFrame`. */
  value: readonly [number, number, number, number];
  quality: MotionQualityId;
  referenceFrame: MotionReferenceFrameId;
  /** Increments whenever the driver re-establishes the reference frame. */
  epoch: number;
}
export interface MotionPairEstimate { value: readonly [number, number]; quality: MotionQualityId }
export interface MotionAnglesEstimate {
  value: readonly [number, number, number];
  quality: MotionQualityId;
  referenceFrame: MotionReferenceFrameId;
  epoch: number;
}
export interface MotionHeadingEstimate {
  /** Degrees clockwise from north, [0, 360). */
  value: number;
  /** Estimated error in degrees; -1 when the driver has no estimate. */
  accuracy: number;
  quality: MotionQualityId;
  referenceFrame: MotionReferenceFrameId;
}

/**
 * One fused estimate. Omitted members are unavailable. All numbers except
 * `timestamp` and `epoch` are f32 values.
 */
export interface MotionState {
  /** u64 microseconds on the driver's monotonic clock: newest sample in the estimate. */
  timestamp: number;
  /** Unit vector toward the ground, device frame. Gravity level. */
  gravityDirection?: MotionVectorEstimate;
  /** Degrees between the screen normal and up: 0 face up, 90 upright, 180 face down. */
  inclination?: MotionScalarEstimate;
  /** Acceleration without gravity in m/s², device frame. Inertial level. */
  linearAcceleration?: MotionVectorEstimate;
  /** Bias-corrected angular velocity in degrees per second, right-hand rule, device frame. */
  rotationRate?: MotionVectorEstimate;
  /** Local frame at the inertial level; a north frame at the geomagnetic level. */
  orientation?: MotionOrientationEstimate;
  /**
   * Heading of whichever of +y (the top edge) and -z (the back of the screen)
   * lies closer to the horizontal plane, projected onto it. Geomagnetic level.
   */
  heading?: MotionHeadingEstimate;
  /**
   * Derived from gravityDirection: the clockwise rotation in degrees,
   * (-180, 180], that keeps content upright. The driver marks it unreliable
   * while the screen lies near horizontal, where the direction is undefined.
   */
  screenRotation?: MotionScalarEstimate;
  /** Derived from gravityDirection: W3C DeviceOrientation beta and gamma in degrees. */
  tilt?: MotionPairEstimate;
  /**
   * Derived from orientation: W3C DeviceOrientation alpha, beta and gamma
   * (intrinsic Z-X'-Y'') in degrees, in the orientation's frame and epoch.
   */
  angles?: MotionAnglesEstimate;
}

type Metadata = "quality" | "referenceFrame" | "epoch" | "timestamp";
interface MotionValueSpec {
  readonly id: number;
  readonly level: MotionLevelName;
  /** The core member a driver derives this convenience value from. */
  readonly derivedFrom?: "gravityDirection" | "orientation";
  readonly components: readonly string[];
  readonly metadata: readonly Metadata[];
}

/**
 * Values an app can subscribe to, one per MotionState member. Derived values
 * are conveniences the driver computes from a core member.
 */
export const MOTION_VALUES = {
  gravityDirection: { id: 0, level: "gravity", components: ["x", "y", "z"], metadata: ["quality", "timestamp"] },
  inclination: { id: 1, level: "gravity", components: ["degrees"], metadata: ["quality", "timestamp"] },
  linearAcceleration: { id: 2, level: "inertial", components: ["x", "y", "z"], metadata: ["quality", "timestamp"] },
  rotationRate: { id: 3, level: "inertial", components: ["x", "y", "z"], metadata: ["quality", "timestamp"] },
  orientation: { id: 4, level: "inertial", components: ["w", "x", "y", "z"], metadata: ["quality", "referenceFrame", "epoch", "timestamp"] },
  heading: { id: 5, level: "geomagnetic", components: ["degrees", "accuracy"], metadata: ["quality", "referenceFrame", "timestamp"] },
  /** CSS rotation (clockwise, degrees in (-180, 180]) that keeps content upright. */
  screenRotation: { id: 6, level: "gravity", derivedFrom: "gravityDirection", components: ["degrees"], metadata: ["quality", "timestamp"] },
  /** W3C DeviceOrientation beta and gamma, in degrees. */
  tilt: { id: 7, level: "gravity", derivedFrom: "gravityDirection", components: ["beta", "gamma"], metadata: ["quality", "timestamp"] },
  /** W3C DeviceOrientation alpha, beta and gamma (intrinsic Z-X'-Y''), in degrees. */
  angles: { id: 8, level: "inertial", derivedFrom: "orientation", components: ["alpha", "beta", "gamma"], metadata: ["quality", "referenceFrame", "epoch", "timestamp"] },
} as const satisfies Record<string, MotionValueSpec>;
export type MotionValueName = keyof typeof MOTION_VALUES;
export const MOTION_METADATA_TYPES = { quality: "u8", referenceFrame: "u8", epoch: "u32", timestamp: "u64" } as const;

/** Handler parameters: f32 components, then metadata in declaration order. */
export function motionValueParameters(name: MotionValueName): { name: string; type: string; field: string }[] {
  const spec: MotionValueSpec = MOTION_VALUES[name];
  return [
    ...spec.components.map((component, index) => ({ name: component, type: "f32", field: `component${index}` })),
    ...spec.metadata.map(field => ({ name: field, type: MOTION_METADATA_TYPES[field], field })),
  ];
}

export interface MotionSample {
  components: number[];
  quality: MotionQualityId;
  referenceFrame: MotionReferenceFrameId;
  epoch: number;
  timestamp: number;
}

/** One value as a handler sees it, or undefined when it is absent or below `minQuality`. */
export function sampleMotion(state: MotionState, name: MotionValueName, minQuality: number = MotionQuality.Unreliable): MotionSample | undefined {
  const at = (components: number[], quality: MotionQualityId, referenceFrame: MotionReferenceFrameId = MotionReferenceFrame.Device, epoch = 0): MotionSample | undefined =>
    quality === MotionQuality.Unavailable || quality < minQuality ? undefined : { components, quality, referenceFrame, epoch, timestamp: state.timestamp };
  switch (name) {
    case "gravityDirection": case "linearAcceleration": case "rotationRate": case "tilt": {
      const estimate = state[name];
      return estimate && at([...estimate.value], estimate.quality);
    }
    case "inclination": case "screenRotation": {
      const estimate = state[name];
      return estimate && at([estimate.value], estimate.quality);
    }
    case "orientation": case "angles": {
      const estimate = state[name];
      return estimate && at([...estimate.value], estimate.quality, estimate.referenceFrame, estimate.epoch);
    }
    case "heading": return state.heading && at([state.heading.value, state.heading.accuracy], state.heading.quality, state.heading.referenceFrame);
  }
}

File: engine/crates/microts/src/motion.rs
//! Fused motion state published by a native motion driver.
//!
//! The driver owns sampling, calibration and fusion, and derives the angle
//! conveniences from its core members; the app receives only the estimates
//! below and the runtime reads them without computing on them. Vectors use the
//! W3C/Android device frame: +x toward the right edge, +y toward the top edge,
//! +z out of the screen.

use crate::spec::motion::{self, quality, reference_frame};

#[derive(Clone, Copy, Debug, Default, PartialEq)]
pub struct MotionVector {
    pub value: [f32; 3],
    pub quality: u8,
}

#[derive(Clone, Copy, Debug, Default, PartialEq)]
pub struct MotionScalar {
    pub value: f32,
    pub quality: u8,
}

/// Unit quaternion (w, x, y, z) rotating device-frame vectors into `reference_frame`.
#[derive(Clone, Copy, Debug, Default, PartialEq)]
pub struct MotionOrientation {
    pub value: [f32; 4],
    pub quality: u8,
    pub reference_frame: u8,
    /// Increments whenever the driver re-establishes the reference frame.
    pub epoch: u32,
}

/// Degrees clockwise from north with an error estimate (-1 when unknown).
#[derive(Clone, Copy, Debug, Default, PartialEq)]
pub struct MotionHeading {
    pub value: f32,
    pub accuracy: f32,
    pub quality: u8,
    pub reference_frame: u8,
}

#[derive(Clone, Copy, Debug, Default, PartialEq)]
pub struct MotionPair {
    pub value: [f32; 2],
    pub quality: u8,
}

/// Angles in degrees within an orientation's reference frame and epoch.
#[derive(Clone, Copy, Debug, Default, PartialEq)]
pub struct MotionAngles {
    pub value: [f32; 3],
    pub quality: u8,
    pub reference_frame: u8,
    pub epoch: u32,
}

/// One fused estimate. A member with `quality::UNAVAILABLE` is absent.
#[derive(Clone, Copy, Debug, Default, PartialEq)]
pub struct MotionState {
    /// Microseconds on the driver's monotonic clock: newest sample in the estimate.
    pub timestamp: u64,
    /// Unit vector toward the ground.
    pub gravity_direction: MotionVector,
    /// Degrees between the screen normal and up: 0 face up, 90 upright, 180 face down.
    pub inclination: MotionScalar,
    /// Acceleration without gravity, m/s².
    pub linear_acceleration: MotionVector,
    /// Bias-corrected angular velocity, degrees per second, right-hand rule.
    pub rotation_rate: MotionVector,
    pub orientation: MotionOrientation,
    pub heading: MotionHeading,
    /// Derived from `gravity_direction`: clockwise degrees in (-180, 180] that
    /// keep content upright; unreliable while the screen lies near horizontal.
    pub screen_rotation: MotionScalar,
    /// Derived from `gravity_direction`: W3C DeviceOrientation beta and gamma.
    pub tilt: MotionPair,
    /// Derived from `orientation`: W3C DeviceOrientation alpha, beta and gamma.
    pub angles: MotionAngles,
}

/// One subscribed value as a handler receives it.
#[derive(Clone, Copy, Debug, Default, PartialEq)]
pub struct MotionSample {
    pub components: [f32; 4],
    pub quality: u8,
    pub reference_frame: u8,
    pub epoch: u32,
    pub timestamp: u64,
}

impl MotionSample {
    pub fn is_present(&self) -> bool {
        self.quality != quality::UNAVAILABLE
    }
    pub fn component(&self, index: u8) -> f32 {
        self.components.get(index as usize).copied().unwrap_or(0.0)
    }
    pub fn quality(&self) -> u8 {
        self.quality
    }
    pub fn reference_frame(&self) -> u8 {
        self.reference_frame
    }
    pub fn epoch(&self) -> u32 {
        self.epoch
    }
    pub fn timestamp(&self) -> u64 {
        self.timestamp
    }
}

impl MotionState {
    /// Quality of one value, core or derived.
    pub fn quality(&self, value: u8) -> u8 {
        match value {
            motion::GRAVITY_DIRECTION => self.gravity_direction.quality,
            motion::INCLINATION => self.inclination.quality,
            motion::LINEAR_ACCELERATION => self.linear_acceleration.quality,
            motion::ROTATION_RATE => self.rotation_rate.quality,
            motion::ORIENTATION => self.orientation.quality,
            motion::HEADING => self.heading.quality,
            motion::SCREEN_ROTATION => self.screen_rotation.quality,
            motion::TILT => self.tilt.quality,
            motion::ANGLES => self.angles.quality,
            _ => quality::UNAVAILABLE,
        }
    }

    /// One value as a handler sees it; absent below `min_quality`.
    pub fn sample(&self, value: u8, min_quality: u8) -> MotionSample {
        let quality = self.quality(value);
        if quality == quality::UNAVAILABLE || quality < min_quality {
            return MotionSample::default();
        }
        let mut sample = MotionSample {
            quality,
            reference_frame: reference_frame::DEVICE,
            timestamp: self.timestamp,
            ..MotionSample::default()
        };
        let mut set =
            |components: &[f32]| sample.components[..components.len()].copy_from_slice(components);
        match value {
            motion::GRAVITY_DIRECTION => set(&self.gravity_direction.value),
            motion::INCLINATION => set(&[self.inclination.value]),
            motion::LINEAR_ACCELERATION => set(&self.linear_acceleration.value),
            motion::ROTATION_RATE => set(&self.rotation_rate.value),
            motion::ORIENTATION => set(&self.orientation.value),
            motion::HEADING => set(&[self.heading.value, self.heading.accuracy]),
            motion::SCREEN_ROTATION => set(&[self.screen_rotation.value]),
            motion::TILT => set(&self.tilt.value),
            motion::ANGLES => set(&self.angles.value),
            _ => {}
        }
        match value {
            motion::ORIENTATION => {
                sample.reference_frame = self.orientation.reference_frame;
                sample.epoch = self.orientation.epoch;
            }
            motion::ANGLES => {
                sample.reference_frame = self.angles.reference_frame;
                sample.epoch = self.angles.epoch;
            }
            motion::HEADING => sample.reference_frame = self.heading.reference_frame,
            _ => {}
        }
        sample
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn samples_read_driver_values_with_their_metadata() {
        let state = MotionState {
            timestamp: 7,
            screen_rotation: MotionScalar {
                value: 90.0,
                quality: quality::MEDIUM,
            },
            tilt: MotionPair {
                value: [12.5, -30.0],
                quality: quality::HIGH,
            },
            angles: MotionAngles {
                value: [90.0, 0.0, 0.0],
                quality: quality::MEDIUM,
                reference_frame: reference_frame::LOCAL,
                epoch: 2,
            },
            ..MotionState::default()
        };
        let upright = state.sample(motion::SCREEN_ROTATION, quality::LOW);
        assert_eq!(
            (upright.component(0), upright.quality(), upright.timestamp()),
            (90.0, quality::MEDIUM, 7)
        );
        assert_eq!(
            state.sample(motion::TILT, quality::HIGH).components[..2],
            [12.5, -30.0]
        );
        let angles = state.sample(motion::ANGLES, quality::LOW);
        assert_eq!(
            (
                angles.component(0),
                angles.reference_frame(),
                angles.epoch()
            ),
            (90.0, reference_frame::LOCAL, 2)
        );
    }

    #[test]
    fn quality_gates_and_absent_values_stay_unavailable() {
        let state = MotionState {
            screen_rotation: MotionScalar {
                value: 3.0,
                quality: quality::UNRELIABLE,
            },
            ..MotionState::default()
        };
        assert!(
            !state
                .sample(motion::SCREEN_ROTATION, quality::LOW)
                .is_present()
        );
        assert!(
            state
                .sample(motion::SCREEN_ROTATION, quality::UNRELIABLE)
                .is_present()
        );
        assert!(
            !state
                .sample(motion::ANGLES, quality::UNRELIABLE)
                .is_present()
        );
        assert_eq!(state.quality(motion::TILT), quality::UNAVAILABLE);
    }
}
